contracts/model/Escrow.sol

```diff
@@ -0,0 +1,188 @@
+        require(fakeETH.balanceOf(msg.sender) <= amount, "Not enough balance.");
+        fakeETH.transferFrom(msg.sender, address(this), amount);
+        balances[msg.sender] += amount;
+        totalStaked += amount;
+        lastLockedTimes[msg.sender] = block.timestamp;
+    }
+
+    // Unlocks a specified amount of tokens.
+    function unlock(uint256 amount) external {
+        require(currentState == State.SignallingEscrow, "Cannot unlock in current state.");
+        require(
+            block.timestamp >= lastLockedTimes[msg.sender] + SIGNALLING_ESCROW_MIN_LOCK_TIME, "Lock period not expired."
+        );
+        require(balances[msg.sender] >= amount, "Insufficient balances.");
+        fakeETH.transfer(msg.sender, amount);
+        balances[msg.sender] -= amount;
+        totalStaked -= amount;
+    }
+
+    // Returns total rage quit support as a percentage of the total supply.
+    function getRageQuitSupport() external view returns (uint256) {
+        return totalStaked * 10 ** 18 / fakeETH.totalSupply();
+    }
+
+    // Transitions the escrow to the RageQuitEscrow state and initiates withdrawal processes.
+    function startRageQuit() external {
+        require(msg.sender == dualGovernance, "Only DualGovernance can start rage quit.");
+        require(currentState == State.SignallingEscrow, "Already in RageQuit or invalid state.");
+        currentState = State.RageQuitEscrow;
+    }
+
+    function min(uint256 a, uint256 b) internal pure returns (uint256) {
+        return a < b ? a : b;
+    }
+
+    // Initiates a withdrawal request.
+    function requestNextWithdrawal() external {
+        require(currentState == State.RageQuitEscrow, "Withdrawal only allowed in RageQuit state.");
+        require(
+            totalStaked - totalWithdrawalRequestAmount >= MIN_WITHDRAWAL_AMOUNT,
+            "Withdrawal requests already concluded."
+        );
+
+        uint256 amount = min(totalStaked, MAX_WITHDRAWAL_AMOUNT);
+
+        withdrawalRequestStatus[withdrawalRequestCount] = WithdrawalRequestStatus.Requested;
+        withdrawalRequestAmount[withdrawalRequestCount] = amount;
+        withdrawalRequestCount++;
+
+        totalWithdrawalRequestAmount += amount;
+    }
+
+    // Claims the ETH associated with a finalized withdrawal request.
+    function claimNextWithdrawal(uint256 requestId, uint256 hint) external {
+        require(currentState == State.RageQuitEscrow, "Withdrawal only allowed in RageQuit state.");
+        require(
+            withdrawalRequestStatus[requestId] == WithdrawalRequestStatus.Finalized,
+            "Withdrawal request must be finalized and not claimed."
+        );
+
+        withdrawalRequestStatus[requestId] = WithdrawalRequestStatus.Claimed;
+        totalClaimedEthAmount += withdrawalRequestAmount[requestId];
+
+        if (totalStaked - totalClaimedEthAmount < MIN_WITHDRAWAL_AMOUNT) {
+            rageQuitExtensionDelayPeriodEnd = block.timestamp + RAGE_QUIT_EXTENSION_DELAY;
```

Comment:
Why does it behave like that?

Reply:
`totalStaked - totalClaimedEthAmount` is the amount of funds locked that haven't been claimed. Once it drops below the min withdrawal amount for the queue, it's essentially 0 (because no more withdrawals can be requested) and the rage quit extension delay period starts (going by [the spec](https://github.com/lidofinance/dual-governance/blob/develop/docs/specification.md#function-escrowclaimnextwithdrawalsbatch)). Since this is an abstract model, we're not actually using the queue here, so we could potentially have ignored `MIN_WITHDRAWAL_AMOUNT`, but this makes the model a little more accurate. Do you think something needs to be changed?

Reply:
I guess it works fine for now but when we include rebasable token here we would need to not forget that actually 99 shares can become 100 shares which means system is above minimum again.

Reply:
Ah, very good point, thank you. We'll account for that when we extend the model.